<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Automated ingestion and conversion of various single-cell data formats. • scKirby</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Automated ingestion and conversion of various single-cell data formats.">
<meta property="og:description" content="Automated ingestion and conversion of various single-cell data formats.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">scKirby</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/conda.html">conda</a>
    </li>
    <li>
      <a href="articles/ingest_data.html">ingest_data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/bschilder/scKirby/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div id="sckirby" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#sckirby" class="anchor"></a>scKirby</h1></div>
<h5>
Author: <i>Brian M. Schilder</i>
</h5>
<h5>
Most recent update: <i>Jun-30-2021</i>
</h5>
<div id="automated-ingestion-and-conversion-of-various-single-cell-data-formats" class="section level2">
<h2 class="hasAnchor">
<a href="#automated-ingestion-and-conversion-of-various-single-cell-data-formats" class="anchor"></a>Automated ingestion and conversion of various single-cell data formats.</h2>
<p>There’s a lot of single-cell omics file/object formats out there, and not all tools support all of these formats. <code>scKirby</code> aims to make switching between these formats much easier by running several steps within a single function: <code><a href="reference/ingest_data.html">ingest_data()</a></code>. Alternatively, users can run any of these steps separately using the designated sub-functions.</p>
<ol>
<li>
<strong>Read</strong>: Automatically infers the file/object type and loads it (sub-function: <code>read_data()</code>).<br>
</li>
<li>
<strong>Convert</strong>: Converts it to the desired file/object type (sub-function: <code>to_&lt;format&gt;</code>).</li>
<li>
<strong>Save</strong>: Saves the converted file/object (sub-function: <code>save_data()</code>).</li>
</ol>
</div>
<div id="documentation-website" class="section level2">
<h2 class="hasAnchor">
<a href="#documentation-website" class="anchor"></a><a href="https://bschilder.github.io/scKirby">Documentation website</a>
</h2>
</div>
<div id="vignette-data-ingestion" class="section level2">
<h2 class="hasAnchor">
<a href="#vignette-data-ingestion" class="anchor"></a><a href="https://bschilder.github.io/scKirby/articles/ingest_data.html">Vignette: data ingestion</a>
</h2>
</div>
<div id="vignette-conda-environments" class="section level2">
<h2 class="hasAnchor">
<a href="#vignette-conda-environments" class="anchor"></a><a href="https://bschilder.github.io/scKirby/articles/conda.html">Vignette: conda environments</a>
</h2>
<p><img src="./images/buff_kirby.jpeg" height="400"></p>
</div>
</div>
<div id="io-formats" class="section level1">
<h1 class="hasAnchor">
<a href="#io-formats" class="anchor"></a>i/o formats</h1>
<div id="supported-input-formats" class="section level2">
<h2 class="hasAnchor">
<a href="#supported-input-formats" class="anchor"></a>Supported input formats</h2>
<ul>
<li><a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html">SingleCellExperiment</a></li>
<li>
<a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">SummarizedExperiment</a><br>
</li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/HDF5Array.html">HDF5SummarizedExperiment</a></li>
<li>
<a href="https://satijalab.org/seurat/index.html">Seurat</a><br>
</li>
<li><a href="https://mojaveazure.github.io/seurat-disk/articles/convert-anndata.html">H5Seurat</a></li>
<li><a href="https://github.com/rcannood/anndata">anndata</a></li>
<li><a href="http://loompy.org/">loom</a></li>
<li><a href="https://satijalab.org/loomR/loomR_tutorial.html">loomR</a></li>
<li><a href="https://github.com/NathanSkene/EWCE">list</a></li>
<li><a href="https://github.com/NathanSkene/EWCE">EWCE</a></li>
<li><a href="https://cran.r-project.org/web/packages/Matrix/index.html">matrix</a></li>
<li><a href="https://slowkow.com/notes/sparse-matrix/">sparseMatrix (dgTMatrix/dgCMatrix)</a></li>
<li><a href="https://petehaitch.github.io/BioC2020_DelayedArray_workshop/articles/Effectively_using_the_DelayedArray_framework_for_users.html">DelayedArray</a></li>
</ul>
</div>
<div id="supported-output-formats" class="section level2">
<h2 class="hasAnchor">
<a href="#supported-output-formats" class="anchor"></a>Supported output formats</h2>
<ul>
<li>
<a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html">SingleCellExperiment</a><br>
</li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">SummarizedExperiment</a></li>
<li>
<a href="https://bioconductor.org/packages/release/bioc/html/HDF5Array.html">HDF5SummarizedExperiment</a><br>
</li>
<li>
<a href="https://satijalab.org/seurat/index.html">Seurat</a><br>
</li>
<li><a href="https://mojaveazure.github.io/seurat-disk/articles/convert-anndata.html">H5Seurat</a></li>
</ul>
</div>
<div id="planned-output-formats" class="section level2">
<h2 class="hasAnchor">
<a href="#planned-output-formats" class="anchor"></a>Planned output formats</h2>
<ul>
<li><a href="https://github.com/rcannood/anndata">anndata</a></li>
<li><a href="http://loompy.org/">loom</a></li>
</ul>
<p><strong>Notes</strong>:</p>
<ul>
<li>For exporting to additional formats, see these following packages:
<ul>
<li>
<a href="https://github.com/cellgeni/sceasy">sceasy</a><br>
</li>
<li><a href="https://theislab.github.io/zellkonverter/articles/zellkonverter.html">zellkonverter</a></li>
</ul>
</li>
<li>Currently, some (but not all) conversions carry over:
<ul>
<li>Multiple assays per experiment.</li>
<li>Additional objects like dimensionality reduction projections (e.g. PCA, tSNE, UMAP) or graphs (e.g. K-nearest neighbors).</li>
</ul>
</li>
</ul>
</div>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"remotes"</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span><span class="op">}</span>

<span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"bschilder/scKirby"</span><span class="op">)</span></code></pre></div>
</div>
<div id="quick-examples" class="section level1">
<h1 class="hasAnchor">
<a href="#quick-examples" class="anchor"></a>Quick examples</h1>
<p>Here are several quick examples of how one can use <code>scKirby</code>. For a complete list of examples please see the <a href="https://bschilder.github.io/scKirby">documentation website</a>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bschilder/scKirby">scKirby</a></span><span class="op">)</span></code></pre></div>
<div id="ingest-list" class="section level2">
<h2 class="hasAnchor">
<a href="#ingest-list" class="anchor"></a>Ingest list</h2>
<p><code>scKirby</code> can ingest a named list (i.e. <code><a href="https://rdrr.io/r/base/list.html">list(exp=..., annot=...)</a></code>) with the following items:</p>
<ul>
<li><p><code>exp</code>: Expression matrix with <em>rows/genes x cols/cells</em>. Can be a variety of matrix classes, including dense or sparse.</p></li>
<li><p><code>annot</code>: Cell annotation <code>data.frame</code> with one cell per row. <code><a href="https://rdrr.io/r/base/colnames.html">rownames(annot)</a></code> should be the same as <code><a href="https://rdrr.io/r/base/colnames.html">colnames(exp)</a></code>.</p></li>
</ul>
<p>This happens to be the format that the example data in <a href="https://github.com/NathanSkene/EWCE"><code>EWCE</code></a> uses, but any user-supplied data will work.</p>
<div id="as-singlecellexperiment" class="section level3">
<h3 class="hasAnchor">
<a href="#as-singlecellexperiment" class="anchor"></a>As <code>SingleCellExperiment</code>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"example_EWCElist"</span><span class="op">)</span>

<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ingest_data.html">ingest_data</a></span><span class="op">(</span>obj<span class="op">=</span><span class="va">example_EWCElist</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="do">## + Returning object directly...</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Converting formats:</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="do">## + 10 core(s) assigned as workers (2 reserved).</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="do">## + EWCElist ==&gt; SingleCellExperiment</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "+ Checking SCE rownames."</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a><span class="do">## + Saving SingleCellExperiment: /var/folders/zq/h7mtybc533b1qzkys_ttgpth0000gn/T//RtmplQASgm/scKirby_output.rds</span></span></code></pre></div>
</div>
<div id="as-seurat" class="section level3">
<h3 class="hasAnchor">
<a href="#as-seurat" class="anchor"></a>As <code>Seurat</code>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ingest_data.html">ingest_data</a></span><span class="op">(</span>obj<span class="op">=</span><span class="va">example_EWCElist</span>, 
                      output_type <span class="op">=</span> <span class="st">"Seurat"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="do">## + Returning object directly...</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Converting formats:</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="do">## + 10 core(s) assigned as workers (2 reserved).</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="do">## + EWCElist ==&gt; Seurat</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="do">## Warning: Feature names cannot have underscores ('_'), replacing with dashes</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a><span class="do">## ('-')</span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a><span class="do">## + Saving Seurat: /var/folders/zq/h7mtybc533b1qzkys_ttgpth0000gn/T//RtmplQASgm/scKirby_output.rds</span></span></code></pre></div>
</div>
</div>
<div id="ingest-seurat" class="section level2">
<h2 class="hasAnchor">
<a href="#ingest-seurat" class="anchor"></a>Ingest Seurat</h2>
<div id="as-singlecellexperiment-1" class="section level3">
<h3 class="hasAnchor">
<a href="#as-singlecellexperiment-1" class="anchor"></a>As <code>SingleCellExperiment</code>
</h3>
<p>In-memory</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"example_seurat"</span><span class="op">)</span>

<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ingest_data.html">ingest_data</a></span><span class="op">(</span>obj<span class="op">=</span><span class="va">example_seurat</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="do">## + Returning object directly...</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Converting formats:</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="do">## + 10 core(s) assigned as workers (2 reserved).</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="do">## + Seurat ==&gt; SingleCellExperiment</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] "+ Checking SCE rownames."</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a><span class="do">## + Saving SingleCellExperiment: /var/folders/zq/h7mtybc533b1qzkys_ttgpth0000gn/T//RtmplQASgm/scKirby_output.rds</span></span></code></pre></div>
</div>
</div>
</div>
<div id="conda-environments" class="section level1">
<h1 class="hasAnchor">
<a href="#conda-environments" class="anchor"></a>Conda environments</h1>
<div id="updating-seurat-objects" class="section level2">
<h2 class="hasAnchor">
<a href="#updating-seurat-objects" class="anchor"></a>Updating Seurat objects</h2>
<p>Seurat’s <code><a href="https://satijalab.org/seurat/reference/UpdateSeuratObject.html">UpdateSeuratObject()</a></code> can only update objects from the version immediately previous to the version of Seurat you currently have installed (e.g. Seurat v2 –&gt; v3). This means you can’t import an object created in Seurat v1 and directly upgrade it to Seurat v3. We have provided yaml files when can be used to create separate envs for each version of Seurat <a href="https://github.com/bschilder/scKirby/tree/main/inst/conda">here</a>.</p>
<p>For more details, see the <a href="https://bschilder.github.io/scKirby/articles/conda.html">scKirby conda env tutorial</a>.</p>
</div>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h1>
<details><div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="do">## R version 4.1.0 (2021-05-18)</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="do">## Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Running under: macOS Big Sur 10.16</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="do">## Matrix products: default</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="do">## BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.dylib</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="do">## LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="do">## locale:</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8</span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a><span class="do">## attached base packages:</span></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a><span class="do">## other attached packages:</span></span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] scKirby_0.1.0</span></span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a><span class="do">## loaded via a namespace (and not attached):</span></span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a><span class="do">##   [1] Seurat_4.0.3                Rtsne_0.15                 </span></span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a><span class="do">##   [3] colorspace_2.0-2            deldir_0.2-10              </span></span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a><span class="do">##   [5] ellipsis_0.3.2              class_7.3-19               </span></span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a><span class="do">##   [7] ggridges_0.5.3              XVector_0.32.0             </span></span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a><span class="do">##   [9] GenomicRanges_1.44.0        gld_2.6.2                  </span></span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a><span class="do">##  [11] spatstat.data_2.1-0         rstudioapi_0.13            </span></span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a><span class="do">##  [13] proxy_0.4-26                leiden_0.3.8               </span></span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a><span class="do">##  [15] listenv_0.8.0               ggrepel_0.9.1              </span></span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a><span class="do">##  [17] fansi_0.5.0                 mvtnorm_1.1-2              </span></span>
<span id="cb10-28"><a href="#cb10-28" aria-hidden="true" tabindex="-1"></a><span class="do">##  [19] codetools_0.2-18            splines_4.1.0              </span></span>
<span id="cb10-29"><a href="#cb10-29" aria-hidden="true" tabindex="-1"></a><span class="do">##  [21] rootSolve_1.8.2.1           knitr_1.33                 </span></span>
<span id="cb10-30"><a href="#cb10-30" aria-hidden="true" tabindex="-1"></a><span class="do">##  [23] polyclip_1.10-0             jsonlite_1.7.2             </span></span>
<span id="cb10-31"><a href="#cb10-31" aria-hidden="true" tabindex="-1"></a><span class="do">##  [25] ica_1.0-2                   cluster_2.1.2              </span></span>
<span id="cb10-32"><a href="#cb10-32" aria-hidden="true" tabindex="-1"></a><span class="do">##  [27] png_0.1-7                   uwot_0.1.10                </span></span>
<span id="cb10-33"><a href="#cb10-33" aria-hidden="true" tabindex="-1"></a><span class="do">##  [29] spatstat.sparse_2.0-0       sctransform_0.3.2          </span></span>
<span id="cb10-34"><a href="#cb10-34" aria-hidden="true" tabindex="-1"></a><span class="do">##  [31] shiny_1.6.0                 compiler_4.1.0             </span></span>
<span id="cb10-35"><a href="#cb10-35" aria-hidden="true" tabindex="-1"></a><span class="do">##  [33] httr_1.4.2                  lazyeval_0.2.2             </span></span>
<span id="cb10-36"><a href="#cb10-36" aria-hidden="true" tabindex="-1"></a><span class="do">##  [35] assertthat_0.2.1            SeuratObject_4.0.2         </span></span>
<span id="cb10-37"><a href="#cb10-37" aria-hidden="true" tabindex="-1"></a><span class="do">##  [37] Matrix_1.3-4                fastmap_1.1.0              </span></span>
<span id="cb10-38"><a href="#cb10-38" aria-hidden="true" tabindex="-1"></a><span class="do">##  [39] later_1.2.0                 htmltools_0.5.1.1          </span></span>
<span id="cb10-39"><a href="#cb10-39" aria-hidden="true" tabindex="-1"></a><span class="do">##  [41] tools_4.1.0                 igraph_1.2.6               </span></span>
<span id="cb10-40"><a href="#cb10-40" aria-hidden="true" tabindex="-1"></a><span class="do">##  [43] gtable_0.3.0                glue_1.4.2                 </span></span>
<span id="cb10-41"><a href="#cb10-41" aria-hidden="true" tabindex="-1"></a><span class="do">##  [45] lmom_2.8                    GenomeInfoDbData_1.2.6     </span></span>
<span id="cb10-42"><a href="#cb10-42" aria-hidden="true" tabindex="-1"></a><span class="do">##  [47] reshape2_1.4.4              RANN_2.6.1                 </span></span>
<span id="cb10-43"><a href="#cb10-43" aria-hidden="true" tabindex="-1"></a><span class="do">##  [49] dplyr_1.0.7                 Rcpp_1.0.6                 </span></span>
<span id="cb10-44"><a href="#cb10-44" aria-hidden="true" tabindex="-1"></a><span class="do">##  [51] scattermore_0.7             Biobase_2.52.0             </span></span>
<span id="cb10-45"><a href="#cb10-45" aria-hidden="true" tabindex="-1"></a><span class="do">##  [53] vctrs_0.3.8                 nlme_3.1-152               </span></span>
<span id="cb10-46"><a href="#cb10-46" aria-hidden="true" tabindex="-1"></a><span class="do">##  [55] lmtest_0.9-38               xfun_0.24                  </span></span>
<span id="cb10-47"><a href="#cb10-47" aria-hidden="true" tabindex="-1"></a><span class="do">##  [57] stringr_1.4.0               globals_0.14.0             </span></span>
<span id="cb10-48"><a href="#cb10-48" aria-hidden="true" tabindex="-1"></a><span class="do">##  [59] mime_0.11                   miniUI_0.1.1.1             </span></span>
<span id="cb10-49"><a href="#cb10-49" aria-hidden="true" tabindex="-1"></a><span class="do">##  [61] lifecycle_1.0.0             irlba_2.3.3                </span></span>
<span id="cb10-50"><a href="#cb10-50" aria-hidden="true" tabindex="-1"></a><span class="do">##  [63] goftest_1.2-2               future_1.21.0              </span></span>
<span id="cb10-51"><a href="#cb10-51" aria-hidden="true" tabindex="-1"></a><span class="do">##  [65] zlibbioc_1.38.0             MASS_7.3-54                </span></span>
<span id="cb10-52"><a href="#cb10-52" aria-hidden="true" tabindex="-1"></a><span class="do">##  [67] zoo_1.8-9                   scales_1.1.1               </span></span>
<span id="cb10-53"><a href="#cb10-53" aria-hidden="true" tabindex="-1"></a><span class="do">##  [69] spatstat.core_2.2-0         spatstat.utils_2.2-0       </span></span>
<span id="cb10-54"><a href="#cb10-54" aria-hidden="true" tabindex="-1"></a><span class="do">##  [71] promises_1.2.0.1            MatrixGenerics_1.4.0       </span></span>
<span id="cb10-55"><a href="#cb10-55" aria-hidden="true" tabindex="-1"></a><span class="do">##  [73] parallel_4.1.0              SummarizedExperiment_1.22.0</span></span>
<span id="cb10-56"><a href="#cb10-56" aria-hidden="true" tabindex="-1"></a><span class="do">##  [75] expm_0.999-6                RColorBrewer_1.1-2         </span></span>
<span id="cb10-57"><a href="#cb10-57" aria-hidden="true" tabindex="-1"></a><span class="do">##  [77] SingleCellExperiment_1.14.1 yaml_2.2.1                 </span></span>
<span id="cb10-58"><a href="#cb10-58" aria-hidden="true" tabindex="-1"></a><span class="do">##  [79] Exact_2.1                   gridExtra_2.3              </span></span>
<span id="cb10-59"><a href="#cb10-59" aria-hidden="true" tabindex="-1"></a><span class="do">##  [81] pbapply_1.4-3               reticulate_1.20            </span></span>
<span id="cb10-60"><a href="#cb10-60" aria-hidden="true" tabindex="-1"></a><span class="do">##  [83] ggplot2_3.3.5               rpart_4.1-15               </span></span>
<span id="cb10-61"><a href="#cb10-61" aria-hidden="true" tabindex="-1"></a><span class="do">##  [85] stringi_1.6.2               S4Vectors_0.30.0           </span></span>
<span id="cb10-62"><a href="#cb10-62" aria-hidden="true" tabindex="-1"></a><span class="do">##  [87] e1071_1.7-7                 BiocGenerics_0.38.0        </span></span>
<span id="cb10-63"><a href="#cb10-63" aria-hidden="true" tabindex="-1"></a><span class="do">##  [89] boot_1.3-28                 BiocParallel_1.26.0        </span></span>
<span id="cb10-64"><a href="#cb10-64" aria-hidden="true" tabindex="-1"></a><span class="do">##  [91] GenomeInfoDb_1.28.0         rlang_0.4.11               </span></span>
<span id="cb10-65"><a href="#cb10-65" aria-hidden="true" tabindex="-1"></a><span class="do">##  [93] pkgconfig_2.0.3             matrixStats_0.59.0         </span></span>
<span id="cb10-66"><a href="#cb10-66" aria-hidden="true" tabindex="-1"></a><span class="do">##  [95] bitops_1.0-7                evaluate_0.14              </span></span>
<span id="cb10-67"><a href="#cb10-67" aria-hidden="true" tabindex="-1"></a><span class="do">##  [97] lattice_0.20-44             tensor_1.5                 </span></span>
<span id="cb10-68"><a href="#cb10-68" aria-hidden="true" tabindex="-1"></a><span class="do">##  [99] ROCR_1.0-11                 purrr_0.3.4                </span></span>
<span id="cb10-69"><a href="#cb10-69" aria-hidden="true" tabindex="-1"></a><span class="do">## [101] htmlwidgets_1.5.3           patchwork_1.1.1            </span></span>
<span id="cb10-70"><a href="#cb10-70" aria-hidden="true" tabindex="-1"></a><span class="do">## [103] cowplot_1.1.1               tidyselect_1.1.1           </span></span>
<span id="cb10-71"><a href="#cb10-71" aria-hidden="true" tabindex="-1"></a><span class="do">## [105] parallelly_1.26.0           RcppAnnoy_0.0.18           </span></span>
<span id="cb10-72"><a href="#cb10-72" aria-hidden="true" tabindex="-1"></a><span class="do">## [107] plyr_1.8.6                  magrittr_2.0.1             </span></span>
<span id="cb10-73"><a href="#cb10-73" aria-hidden="true" tabindex="-1"></a><span class="do">## [109] R6_2.5.0                    IRanges_2.26.0             </span></span>
<span id="cb10-74"><a href="#cb10-74" aria-hidden="true" tabindex="-1"></a><span class="do">## [111] DescTools_0.99.42           generics_0.1.0             </span></span>
<span id="cb10-75"><a href="#cb10-75" aria-hidden="true" tabindex="-1"></a><span class="do">## [113] DelayedArray_0.18.0         DBI_1.1.1                  </span></span>
<span id="cb10-76"><a href="#cb10-76" aria-hidden="true" tabindex="-1"></a><span class="do">## [115] mgcv_1.8-36                 pillar_1.6.1               </span></span>
<span id="cb10-77"><a href="#cb10-77" aria-hidden="true" tabindex="-1"></a><span class="do">## [117] fitdistrplus_1.1-5          abind_1.4-5                </span></span>
<span id="cb10-78"><a href="#cb10-78" aria-hidden="true" tabindex="-1"></a><span class="do">## [119] survival_3.2-11             RCurl_1.98-1.3             </span></span>
<span id="cb10-79"><a href="#cb10-79" aria-hidden="true" tabindex="-1"></a><span class="do">## [121] tibble_3.1.2                future.apply_1.7.0         </span></span>
<span id="cb10-80"><a href="#cb10-80" aria-hidden="true" tabindex="-1"></a><span class="do">## [123] crayon_1.4.1                KernSmooth_2.23-20         </span></span>
<span id="cb10-81"><a href="#cb10-81" aria-hidden="true" tabindex="-1"></a><span class="do">## [125] utf8_1.2.1                  spatstat.geom_2.2-0        </span></span>
<span id="cb10-82"><a href="#cb10-82" aria-hidden="true" tabindex="-1"></a><span class="do">## [127] plotly_4.9.4.1              rmarkdown_2.9              </span></span>
<span id="cb10-83"><a href="#cb10-83" aria-hidden="true" tabindex="-1"></a><span class="do">## [129] grid_4.1.0                  data.table_1.14.0          </span></span>
<span id="cb10-84"><a href="#cb10-84" aria-hidden="true" tabindex="-1"></a><span class="do">## [131] digest_0.6.27               xtable_1.8-4               </span></span>
<span id="cb10-85"><a href="#cb10-85" aria-hidden="true" tabindex="-1"></a><span class="do">## [133] tidyr_1.1.3                 httpuv_1.6.1               </span></span>
<span id="cb10-86"><a href="#cb10-86" aria-hidden="true" tabindex="-1"></a><span class="do">## [135] stats4_4.1.0                munsell_0.5.0              </span></span>
<span id="cb10-87"><a href="#cb10-87" aria-hidden="true" tabindex="-1"></a><span class="do">## [137] viridisLite_0.4.0</span></span></code></pre></div>
</details>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/bschilder/scKirby/">https://​github.com/​bschilder/​scKirby/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/bschilder/scKirby/issues">https://​github.com/​bschilder/​scKirby/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Brian Schilder <br><small class="roles"> Maintainer </small> <a href="https://orcid.org/0000-0001-5949-2191" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Brian Schilder.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
